File: src/main/java/br/com/app/estoque/domain/service/TransferenciaService.java
package br.com.app.estoque.domain.service;

import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import br.com.app.estoque.domain.enuns.StatusTransferencia;
import br.com.app.estoque.domain.model.Transferencia;
import br.com.app.estoque.domain.repository.TransferenciaRepository;
import br.com.app.infraestructure.exception.MensagemException;
import br.com.app.infraestructure.security.SessionUsuario;

@Service
@Transactional(readOnly = true, propagation = Propagation.REQUIRED)
public class TransferenciaService {

	@Autowired
	private TransferenciaRepository transferenciaRepository;
	@Autowired
	private BaixaEstoqueService baixarEstoque;
	@Autowired
	private EntradaEstoqueService entradaEstoque;
	 
	@Transactional(readOnly = false)
	public void salvarAlterar(Transferencia transferencia) {

		transferencia.novaTransferencia();
		baixarEstoque.baixar(transferencia);
		transferenciaRepository.save(transferencia);
	}

	@Transactional(readOnly = false)
	public void aceitarTransferencia(Long numeroNota) {
		Long idEmpreendimento = SessionUsuario.getInstance().getUsuario().getEmpreendimento().getId();
		Transferencia transferencia = transferenciaRepository.buscarTransferenciaPorNumeroNota(numeroNota);
		
		if(!idEmpreendimento.equals(transferencia.getEmpreendimentoDestino().getId())) {
			throw new MensagemException("Somente o Empreendimento " + transferencia.getEmpreendimentoDestino().getDescricao() + " pode aceitar a transferencia" );
		}
		transferencia.setStatusTransferencia(StatusTransferencia.EFETUADO);	
		entradaEstoque.entradaEstoque(transferencia);
		transferenciaRepository.save(transferencia);
		

	}

	@Transactional(readOnly = false)
	public void rejeitarTransferencia(Long numeroNota) {
		Long idEmpreendimento = SessionUsuario.getInstance().getUsuario().getEmpreendimento().getId();
		Transferencia transferencia = transferenciaRepository.buscarTransferenciaPorNumeroNota(numeroNota);
		
		if(!idEmpreendimento.equals(transferencia.getEmpreendimentoDestino().getId())) {
			throw new MensagemException("Somente o Empreendimento " + transferencia.getEmpreendimentoDestino().getDescricao() + " pode rejeitar a transferencia" );
		}		
		transferencia.rejeitarTransferencia();
		CancelamentoTransferencia cancelamento = new CancelamentoTransferencia(transferencia);
		entradaEstoque.entradaEstoque(cancelamento);
		transferenciaRepository.save(transferencia);
	}

	
	public Optional<Transferencia> buscaPorId(Long id) {
		return transferenciaRepository.findById(id);
	}

	

	public Page<Transferencia> findAll(String tipo, Pageable page) {
		Long idEmpreendimento = SessionUsuario.getInstance().getUsuario().getEmpreendimento().getId();
		if(tipo.equalsIgnoreCase("ENVIADOS")) {
			return	transferenciaRepository.findByNotaFiscalEmpreendimentoId(idEmpreendimento, page);
		}else {
			return transferenciaRepository.findByEmpreendimentoDestinoId(idEmpreendimento, page);
		}		
	}

	public Page<Transferencia> findByCodigo(String tipo,String descricao, Pageable page) {
		Long idEmpreendimento = SessionUsuario.getInstance().getUsuario().getEmpreendimento().getId();
		Page<Transferencia> estoques = null;
		descricao = descricao.replaceAll("[./-]", "");

		if(tipo.equalsIgnoreCase("ENVIADOS")) {
			estoques = transferenciaRepository.findByNotaFiscalNumeroAndNotaFiscalEmpreendimentoId(Long.valueOf(descricao),
					idEmpreendimento, page);
		}else {
			estoques = transferenciaRepository.findByNotaFiscalNumeroAndEmpreendimentoDestinoId(Long.valueOf(descricao),
					idEmpreendimento, page);
		}	
		
			if(estoques.getNumberOfElements() < 1) {
				throw new MensagemException("Nota fiscal não encontrada, com o código: " +  descricao);
			}
			
		return estoques;
	}
}

File: bin/src/main/resources/static/views/pages/modulo_estoque/transferencia/js/transferencia-service.js
app.factory('TransferenciaService', function($rootScope, toastr, $http,$q){
	
	var url = '/rest/almoxarifado/transferencia';
	return{
		insert: function(obj){
			return $http.post(url, obj)
			.then(function(response){
				return response.data;
			},function(errResponse){
					return $q.reject(errResponse);
			});
		},
		
		findByTextAndPagination: function(tipo , text , page ,  orderBy, direction , maxResults){
			var config = {params: {descricao : text, page: page , maxResults : maxResults, orderBy : orderBy, direction : direction, tipo : tipo}};
			return $http.get(url + '/buscar', config)
			.then(function(response){
				return response.data;
			},function(errResponse){
				return $q.reject(errResponse);
			});
		},
		
		findById: function(param){
			return $http.get(url + '/'+param)
			.then(function(response){
				return response.data;
			},function(errResponse){
				return $q.reject(errResponse);
			});
		},
		aceitarTransferencia: function(numeroNota){
			return $http.post(url + '/aceitar', numeroNota)
			.then(function(response){
				return response.data;
			},function(errResponse){
				return $q.reject(errResponse);
			});
		},		
		rejeitarTransferencia: function(numeroNota){
			return $http.post(url + '/rejeitar', numeroNota)
			.then(function(response){
				return response.data;
			},function(errResponse){
				return $q.reject(errResponse);
			});
		},		
		
		
	}
});